asmcomp/cmm_helpers.ml

```diff
@@ -1410,24 +1513,40 @@ let make_switch arg cases actions dbg =
     Cmmgen_state.add_constant table (Const_table (Local,
         Array.to_list (Array.map (fun act ->
           const_actions.(act)) cases)));
-    addr_array_ref (Cconst_symbol (table, dbg)) (tag_int arg dbg) dbg
+    (* Constant integers loaded from a table are tagged,
+       so that Cload never produces untagged integers. *)
+    addr_array_ref (Cconst_symbol (table, dbg)) arg dbg
   in
   let make_affine_computation ~offset ~slope arg dbg =
-    (* In case the resulting integers are an affine function of the index, we
-       don't emit a table, and just compute the result directly *)
-    add_int
-      (mul_int arg (natint_const_untagged dbg slope) dbg)
-      (natint_const_untagged dbg offset)
-      dbg
+    add_int_caml
+      (mul_int_caml (int_const dbg slope) arg dbg)
+      (int_const dbg offset) dbg
   in
   match Misc.Stdlib.Array.all_somes (Array.map extract_uconstant actions) with
   | None ->
-      Cswitch (arg,cases,actions,dbg)
+      let arg = match arg with
+        | Tagged arg -> untag_int arg dbg
+        | Untagged arg -> arg
+      in
```

Comment:
Please move all these switches earlier:
```ocaml
let tagged_arg, untagged_arg =
  match arg with 
  | Tagged arg -> arg, untag_int arg dbg
  | Untagged arg -> tag_int arg dbg, arg
in
```

Reply:
If you want to check that the new code is equivalent, in terms of amount-of-tagging, to the previous implementation, I suspect that it's easier with the current approach, because the reasoning is different in each case. If you are already convinced and prefer to see shorter code, I don't mind using the refactoring you suggest. I could also use `arg_{tagged,untagged}` as a naming convention for the fixed-taggedness arguments of the auxiliary functions, which you suggest documenting better below.

Reply:
I did something along these lines in a new commit, see if you like it.

Reply:
Checking that the code was equivalent was already complicated because some functions that expected untagged values now expect tagged values. The pattern-matches were annoying because all I want to know is whether we're using the untagged or tagged version at a given place. It's much more obvious with a suitably named variable than with pattern-matching.
The naming convention change would be really nice; I will ask our local Cmm specialist (@Gbury) if he wants to help with that (a lot of functions in this file could benefit from it).